feat: IFS-1802 - Vorgaben zu bidirektionalen Assoziationen (#694)
IsyFact/IsyFact-Standards: isyfact-standards-doc/src/docs/antora/modules/isy-persistence/pages/konzept/vorgaben-konventionen.adoc

Das identifizierende Attribut darf beliebige Typen besitzen: Es dürfen, neben numerischen Werten, auch Zeichenketten oder Datumsangaben sein.

[[definition-von-assoziationen]]
=== Definition von Assoziationen

[[eins-zu-n-assoziationen]]
==== 1-zu-n Assoziationen

Eine 1-zu-n-Assoziation (siehe https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#collections[Collection Mapping]) ist in der Regel als unsortierte Menge (`Set`) zu definieren, da in dieser keine Reihenfolge definiert ist.
Wird von der Anwendung eine Sortierung benötigt und sind alle für die Sortierung benötigten Attribute in der Entität enthalten, dann kann auch eine Liste (`List`) verwendet werden, da die Datenbank effizienter sortieren kann als eine Java-Implementierung.

[[bidirektionale-assoziationen]]
==== Bidirektionale Assoziationen

> Bidirektionale Assoziation beschreibt die Beziehung zwischen zwei Entitäten, wobei jede Entität einen Verweis auf die andere Entität hat. Sie ermöglicht es Ihnen, von einer Entität zu einer anderen Entität zu navigieren, die mit ihr verbunden ist, und umgekehrt.

Reviewer: ```suggestion
Bidirektionale Assoziation beschreibt die Beziehung zwischen zwei Entitäten, wobei jede Entität einen Verweis auf die andere Entität besitzt. 
Sie ermöglicht es ihnen, von einer Entität zu einer anderen Entität zu navigieren, die mit ihr verbunden ist, und umgekehrt.
```

Es gibt 4 verschiedene Arten der bidirektionalen Assoziation, die wie folgt aussehen:

. Bidirektionale Eins-zu-Eins-Verknüpfung (One-to-One)
. Bidirektionale Eins-zu-Viele-Verknüpfung (One-to-Many)
. Bidirektionale Viele-zu-Eins-Verknüpfung (Many-to-One)

> . Bidirektionale Viele-zu-Viele-Verknüpfung (Many-to-Many )

Reviewer: ```suggestion
. Bidirektionale Eins-zu-Eins-Verknüpfung (one-to-one)
. Bidirektionale Eins-zu-Viele-Verknüpfung (one-to-many)
. Bidirektionale Viele-zu-Eins-Verknüpfung (many-to-one)
. Bidirektionale Viele-zu-Viele-Verknüpfung (many-to-many)
```

===== https://docs.jboss.org/hibernate/stable/orm/userguide/html_single/Hibernate_User_Guide.html#associations-one-to-many-bidirectional[JPA bidirektionale Sync-Methoden]

> Wenn eine bidirektionale Assoziation gebildet wird, muss der Anwendungsentwickler sicherstellen, dass beide Seiten zu jeder Zeit synchron sind.

Reviewer: ```suggestion
Wenn eine bidirektionale Assoziation gebildet wird, muss sichergestellt werden, dass beide Seiten zu jeder Zeit synchron sind.
```

===== Vermeiden von to-many-Assoziationen

> Bei einer to-many Assoziation lädt Hibernate alle zugehörigen Entitäten, wenn es die Assoziation initialisiert. Das kann mehrere Sekunden oder sogar Minuten dauern, wenn Hibernate mehrere tausend Entitäten abrufen muss.

Reviewer: ```suggestion
Bei einer to-many-Assoziation lädt Hibernate alle zugehörigen Entitäten, wenn es die Assoziation initialisiert. 
Das kann mehrere Sekunden oder sogar Minuten dauern, wenn Hibernate mehrere tausend Entitäten abrufen muss.
```

Ist das so? [Einer der Artikel](https://thorben-janssen.com/best-practices-for-many-to-many-associations-with-hibernate-and-jpa/) besagt, dass Hibernate standardmäßig den `FetchType.LAZY` für diese Art von Assoziation benutzt. Dann würde der Zugriff erst bei der ersten Nutzung erfolgen. Das Problem besteht natürlich trotzdem.

Author: Bei FetchType.LAZY wird tatsächlich später geladen. Es kommt letztendlich auf das Szenario an ob dies möglich ist.  Leider wird Caching und FetchType in einen anderen Abschnitt behandelt. Ich würde es begrüßen wenn man die Betrachtung auf konkrete Szenarien macht

> Stattdessen sollte für die Traversierung in Gegenrichtung (many-to-one) eine Query verwendet werden.

Reviewer: Steht das in einer der Quellen? Mir ist die neue Formulierung leider auch nicht klarer als die alte. Falls ja, bietet sich ein Verweis auf die Quelle an.

Author: Die Aussage dass hier eine Query zu verwenden ist war so auch im vorherigen Text. Die gelöschten Zeilen sind durch die Sync Methoden entfallen. 
Tatsächlich kann man dies auch aus der letzten Quellenangabe ableiten. Wenn ich zwischen den Entitäten Parent und Child eine 1 zu n Beziehung habe und z.B. nach dem ChildName suche, so kann ich dies auch durch eine Query umdrehen und erhalte genau eine Child Entität mit zugehörigen Parent

> ===== Best Practices
> Generell sollten bei der Verwendung von Assoziationen die Best Practices beachtet werden
> 
> ===== Quellen
> * https://docs.jboss.org/hibernate/stable/orm/userguide/html_single/Hibernate_User_Guide.html#associations[Hibernate User Guide - Associations]
> * https://vladmihalcea.com/jpa-bidirectional-sync-methods[Vlad Mihalcea - JPA Best practices]
> * https://thorben-janssen.com/best-practices-for-many-to-many-associations-with-hibernate-and-jpa[Thorben Janssen - JPA Best practices]
> * https://thorben-janssen.com/best-practices-many-one-one-many-associations-mappings[Thorben Janssen - Best practices  many-one-one-many-associations]

Reviewer: ```suggestion
[NOTE] 
====
Die folgenden Best Practices helfen bei der Verwendung von Assoziationen:

* https://docs.jboss.org/hibernate/stable/orm/userguide/html_single/Hibernate_User_Guide.html#associations[Hibernate User Guide - Associations]
* https://vladmihalcea.com/jpa-bidirectional-sync-methods[Vlad Mihalcea - JPA Best practices]
* https://thorben-janssen.com/best-practices-for-many-to-many-associations-with-hibernate-and-jpa[Thorben Janssen - JPA Best practices]
* https://thorben-janssen.com/best-practices-many-one-one-many-associations-mappings[Thorben Janssen - Best practices  many-one-one-many-associations]

====
```
